source/develop/release-management/features/virt/engine-dedicated-cpus.md

```diff
@@ -0,0 +1,69 @@
+---
+Title: Engine Dedicated CPUs
+Category: feature
+Authors: sabusale
+---
+
+# Engine Dedicated CPUs
+
+## Summary
+
+Once choosing dedicated CPU policy, The guest vCPUs will be strictly pinned to a set of host pCPUs (similarly to static CPU pinning). The set of pCPUs will be chosen to match the required guest CPU topology.
+
+## Owner
+
+*   Name: Saif Abu Saleh (sabusale)
+*   Email: sabusale@redhat.com
+
+
+### Tasks:
+
+
+1. Add dedicated to CPU Pinning Policy enum
+
+2.  Scheduling
+
+3. Send vcpupin with cpusets and policy name in VM metadata(only when Policy is dedicated?)
+   1. Send the policy name from vmStatic policy field 
+   2. Send the cpusets (convert from CurrentCpuPinning field form to array)
+
+4. API: Migrate
+   1. Send cpusets additional parameter (convert from CurrentCpuPinning field form to array)
+   
+5. setNumberOfCpus VDSM call
+   1. Send cpusets additional parameter (convert from CurrentCpuPinning field form to array)
+
+
+### Data Model:
+
+* Save CPU mappings in a field (maybe use CurrentCpuPinning from vm_dynamic)?
```

Comment:
I think I am missing few things:

If VDSM handles and have the mappings between vCpu and pCPU:
- why the engine need to send the mapping (vcpupin field in domain XML?)
- why the engine need to send the mappings in VM.migrate ?
- I may be missing something but I don't see in the VDSM that the mapping between vCPU and pCPU is exposed to the engine

The CPU topology doesn't really show the mapping
pinned_cpus property is not really exposed

So how the engine should get the mapping from VDSM?

Reply:
I think the flows goes like this:
1. 2 VDSMs are running and reporting the following CPUs are shared/available/free

- VDSM A:  CPUs 1, 2 on socket 1 and 3, 4 on socket 2
- VDSM B:  CPUs 1 on socket 1 and 4 on socket 2

2. User wants to start VM that requires 2 CPUs and has dedicated cpu policy
3. Scheduller ranks all hosts that have enough physical CPUs available using the selected cpu policy. In our case, both hosts have enough pCPUs, however, the A has both available on one socket whereas the B would need to divide them on 2 different sockets. So A gets higher rank than B
4. Host A is selected to run the VM
5. Engine creates the mapping between CPUs as: 1#1_2#2
6. Engine sends the domain XML to VDSM with the required mapping
7. VDSM updates its shared cpu pool and removes the mapped cpus (possibly with some more based on the selected policy)
8. VDSM runs the VM with pinned CPUs
9. VDSM A reports that only cpus 3,4 are in a shared pool/available

It is the engine that contains all the logic how the cpus should be pinned so that it can select the best host from all available hosts. VDSM only keeps the list of how its resources (CPUs) are utilized.